packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/ImagePickerUtils.kt

```diff
@@ -1,91 +1,170 @@
 import expo.modules.imagepicker.ImagePickerConstants.TAG
+import kotlinx.coroutines.runInterruptible
 import java.io.File
+import java.io.FileNotFoundException
+import java.io.FileOutputStream
 import java.io.IOException
 
-/**
- * Fallback method for getting file type from url string as [ContentResolver.getType] might sometimes return `null`
- */
+internal fun createOutputFile(cacheDir: File, extension: String): File {
+  val filePath = FileUtilities.generateOutputPath(cacheDir, ImagePickerConstants.CACHE_DIR_NAME, extension)
+  return try {
+    File(filePath).apply { createNewFile() }
+  } catch (cause: IOException) {
+    throw FailedToCreateFileException(filePath, cause)
+  }
+}
+
+internal fun getType(contentResolver: ContentResolver, uri: Uri): String =
+  contentResolver.getType(uri)
+    ?: getTypeFromFileUrl(uri.toString())
+    ?: throw FailedToDeduceTypeException()
+
 private fun getTypeFromFileUrl(url: String): String? {
   val extension = MimeTypeMap.getFileExtensionFromUrl(url)
-  return if (extension != null) MimeTypeMap.getSingleton().getMimeTypeFromExtension(extension) else null
+  return extension?.let { MimeTypeMap.getSingleton().getMimeTypeFromExtension(it) }
 }
 
-fun createOutputFile(cacheDir: File, extension: String): File? {
+/**
+ * Convert this [File] to [Uri] that might be accessed by 3rd party Activities, eg. by camera application
+ */
+internal fun File.toContentUri(context: Context): Uri {
   return try {
-    File(generateOutputPath(cacheDir, "ImagePicker", extension))
-      .apply {
-        createNewFile()
-      }
-  } catch (e: IOException) {
-    null
+    FileProvider.getUriForFile(context, context.packageName + ".ImagePickerFileProvider", this)
+  } catch (e: Exception) {
+    Uri.fromFile(this)
   }
 }
 
-fun getType(contentResolver: ContentResolver, uri: Uri): String? {
-  return contentResolver.getType(uri) ?: getTypeFromFileUrl(uri.toString())
+internal fun File.toBitmapCompressFormat(): Bitmap.CompressFormat = when {
+  this.extension.contains("png") -> Bitmap.CompressFormat.PNG
+  else -> Bitmap.CompressFormat.JPEG
 }
 
-fun contentUriFromFile(file: File, application: Application): Uri {
-  return try {
-    FileProvider.getUriForFile(application, application.packageName + ".ImagePickerFileProvider", file)
-  } catch (e: Exception) {
-    Uri.fromFile(file)
+internal fun Bitmap.CompressFormat.toImageFileExtension(): String {
+  return when (this) {
+    Bitmap.CompressFormat.PNG -> ".png"
+    Bitmap.CompressFormat.JPEG -> ".jpeg"
+    else -> throw RuntimeException("Compress format not supported '${this.name}'")
   }
 }
 
-// http://stackoverflow.com/a/38858040/1771921
-fun uriFromFile(file: File): Uri = Uri.fromFile(file)
-
-fun uriFromFilePath(path: String) = uriFromFile(File(path))
+internal fun String.toImageFileExtension(): String = when {
+  this.contains("png") -> ".png"
+  this.contains("gif") -> ".gif"
+  this.contains("bmp") -> ".bmp"
+  !this.contains("jpeg") -> {
```

Comment:
```suggestion
  this.endsWith("png", ignoreCase = true) -> ".png"
  this.endsWith("gif", ignoreCase = true) -> ".gif"
  this.endsWith("bmp", ignoreCase = true) -> ".bmp"
  !this.endsWith("jpeg", ignoreCase = true) -> {
```
I think it should be done this way, cause someone can have a file called `png.gif` 😅